inline_translated.c

```diff
@@ -0,0 +1,554 @@
+    case Instr_Jump:
+        result.length = 2;
+        assert(addr+1 < PROGRAM_SIZE);
+        result.immediate = (int32_t)prog[addr+1];
+        break;
+    case Instr_Break:
+    default: /* Undefined instructions equal to Break */
+        result.length = 1;
+        result.opcode = Instr_Break;
+        break;
+    }
+    return result;
+}
+
+/* Supporting functions for in-lined service routines */
+
+static void enter_generated_code(void* addr) {
+    __asm__ __volatile__ ( "jmp *%0"::"r"(addr):);
+}
+
+static void exit_generated_code() {
+    longjmp(return_buf, 1);
+}
+
+static inline void push(cpu_t *pcpu, uint32_t v) {
+    assert(pcpu);
+    if (pcpu->sp >= STACK_CAPACITY-1) {
+        printf("Stack overflow\n");
+        pcpu->state = Cpu_Break;
+        exit_generated_code();
+    }
+    pcpu->stack[++pcpu->sp] = v;
+}
+
+static void inline_translate_program(const Instr_t *prog,
+                           char *out_code, void **entrypoints, int len) {
+    assert(prog);
+    assert(out_code);
+    assert(entrypoints);
+
+    /* An IA-32 instruction "CALL rel32" is used as a trampoline to invoke
+       service routines. A template for it is "call .+0x00000005" */
+    const char call_template_code[] = { 0xe8, 0x00, 0x00, 0x00, 0x00 };
+    const int call_template_size = sizeof(call_template_code);
+
+    int i = 0; /* Address of current guest instruction */
+    char* cur = out_code; /* Where to put new code */
+
+    while (i < len) {
+        decode_t decoded = decode_at_address(prog, i);
+        entrypoints[i] = (void*) cur;
+
+        int addr_exit1 = 0;
+        int addr_exit2 = 0;
+        int addr_exit3 = 0;
+        int addr_puts1 = 0;
+        int addr_puts2 = 0;
+        int addr_push1 = 0;
+        int addr_push2 = 0;
+        int addr_push3 = 0;
+        int addr_abort = 0;
+        int addr_printf = 0;
+        int addr_rand = 0;
+        int addr_str1 = (intptr_t)&str_printf;
+        int addr_str2 = (intptr_t)&str_pop;
+        int addr_str3 = (intptr_t)&str_push;
+        int len = 0;
+        int imm = 0;
+
+        switch(decoded.opcode) {
+        case Instr_Nop:
+            addr_exit1 = (intptr_t)&exit_generated_code - (intptr_t)cur - call_template_size
+                                - (intptr_t)0x2b;
+            len = sizeof(bin_sr_Nop);
+            assert(cur + len - out_code < JIT_CODE_SIZE);
+            memcpy(cur, bin_sr_Nop, len);
+            memcpy(cur+0x2c,&addr_exit1, 4);
+            break;
+        case Instr_Halt:
+            addr_exit1 = (intptr_t)&exit_generated_code - (intptr_t)cur - call_template_size
+                                - (intptr_t)0x11;
+            len = sizeof(bin_sr_Halt);
+            assert(cur + len - out_code < JIT_CODE_SIZE);
+            memcpy(cur, bin_sr_Halt, len);
+            memcpy(cur+0x12,&addr_exit1, 4);
+            break;
+        case Instr_Print:
+            addr_exit1 = (intptr_t)&exit_generated_code - (intptr_t)cur - call_template_size

```

Comment:
This whole subcase needs a proper commenting, because quite heavy address fiddling is done here.